### PR TITLE
AnimatedCrossFade layout customisation

### PR DESCRIPTION
**Pull request by Hixie**
* Optimise AnimatedSize for the tight case.
* Remove `default` from a switch statement over enum (so that analyzer will complain if we add enum values).
* Adopt the Size since we use it after the child may have changed (which would throw normally).
* AnimatedCrossFade.layoutBuilder

### CODE REVIEW

File: packages/flutter/lib/src/rendering/animated_size.dart [hunk 1]
```diff
@@ -158,12 +158,20 @@ class RenderAnimatedSize extends RenderAligningShiftedBox {
 
     if (child == null) {
       size = _sizeTween.begin = _sizeTween.end = constraints.smallest;
+      _state = RenderAnimatedSizeState.start;
+      return;
```
Comment (yjbanov) [inline packages/flutter/lib/src/rendering/animated_size.dart]: Should we also `_controller.stop()`?

File: packages/flutter/lib/src/rendering/animated_size.dart [hunk 2]
```diff
@@ -158,12 +158,20 @@ class RenderAnimatedSize extends RenderAligningShiftedBox {
...
+    }
+    if (constraints.isTight) {
+      size = _sizeTween.begin = _sizeTween.end = constraints.smallest;
+      child.layout(constraints);
+      _state = RenderAnimatedSizeState.start;
       return;
```
Comment (yjbanov) [inline packages/flutter/lib/src/rendering/animated_size.dart]: `_controller.stop()`?

File: packages/flutter/lib/src/rendering/box.dart [hunk 3]
```diff
@@ -15,7 +15,7 @@ import 'object.dart';
 
 // This class should only be used in debug builds.
 class _DebugSize extends Size {
```
Comment (yjbanov) [inline packages/flutter/lib/src/rendering/box.dart]: unrelated to changes in this PR: I wonder if the existence of this subclass causes the AOT compiler to believe that `Size` is polymorphic and makes it slower than it could be.

@rmacnak-google?

File: packages/flutter/lib/src/rendering/box.dart [hunk 4]
```diff
@@ -1484,20 +1484,74 @@ abstract class RenderBox extends RenderObject {
...
+  Size debugAdoptSize(Size value) {
+    Size result = value;
     assert(() {
-      _size = new _DebugSize(_size, this, debugCanParentUseSize);
+      if (value is _DebugSize) {
+        if (value._owner != this) {
+          if (value._owner.parent != this) {
+            throw new FlutterError(
+              'The size property was assigned a size inappropriately.\n'
+              'The following render object:\n'
```
Comment (yjbanov) [inline packages/flutter/lib/src/rendering/box.dart]: s/object/box/?

File: packages/flutter/lib/src/rendering/box.dart [hunk 5]
```diff
@@ -1484,20 +1484,74 @@ abstract class RenderBox extends RenderObject {
...
+              'A child\'s size was used without setting parentUsesSize.\n'
+              'The following render object:\n'
+              '  $this\n'
+              '...was assigned a size obtained from its child:\n'
+              '  ${value._owner}\n'
+              'However, when the child was laid out, the parentUsesSize argument '
+              'was not set or set to false. Subsequently this transpired to be '
+              'inaccurate: the size was nonetheless used by the parent.\n'
+              'It is important to tell the framework if the size will be used or not '
+              'as several important performance optimisations can be made if the '
```
Comment (yjbanov) [inline packages/flutter/lib/src/rendering/box.dart]: nit: optimizations

File: packages/flutter/lib/src/widgets/animated_cross_fade.dart [hunk 6]
```diff
@@ -24,6 +24,19 @@ enum CrossFadeState {
...
+///
+/// The `topChild` is the child fading in, which is normally drawn on top. The
+/// `bottomChild` is the child fading out, normally drawn on the bottom.
+///
+/// For good performance, the returned widget tree should contain both the
+/// `topChild` and the `bottomChild`; the depth of the tree, and the types of
+/// the widgets in the tree, from the returned widget to each of the children
+/// should be the same; and where there is a widget with multiple children, the
+/// top child and the bottom child should be keyed using the provided
+/// `topChildKey` and `bottomChildKey` keys respectively.
```
Comment (yjbanov) [inline packages/flutter/lib/src/widgets/animated_cross_fade.dart]: I'd add `See [AnimatedCrossFade.defaultLayoutBuilder] for an example implementation'.

I also wonder if the keys should be uniquely generated `GlobalKey`s instead, thus allowing more interesting layouts than those based on multiple children. In fact, I wonder if we would need to pass the keys here at all if they are `GlobalKey`s?

**Review by yjbanov — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/rendering/animated_size.dart (hunk 2 above)
Comment (Hixie) [inline packages/flutter/lib/src/rendering/animated_size.dart]: ah yeah good point

File: packages/flutter/lib/src/widgets/animated_cross_fade.dart (hunk 6 above)
Comment (Hixie) [inline packages/flutter/lib/src/widgets/animated_cross_fade.dart]: We want to avoid using global keys if at all possible. They're expensive.

You can always convert a LocalKey to a GlobalKey if you really need a global key on the receiving end.

File: packages/flutter/lib/src/rendering/animated_size.dart (hunk 1 above)
Comment (Hixie) [inline packages/flutter/lib/src/rendering/animated_size.dart]: done

File: packages/flutter/lib/src/rendering/animated_size.dart (hunk 2 above)
Comment (Hixie) [inline packages/flutter/lib/src/rendering/animated_size.dart]: merged these two blocks together and just made the layout use elvis:

```dart
    if (child == null || constraints.isTight) {                                                                                                                                                                                                             
      _controller.stop();                                                                                                                                                                                                                                   
      size = _sizeTween.begin = _sizeTween.end = constraints.smallest;                                                                                                                                                                                      
      _state = RenderAnimatedSizeState.start;                                                                                                                                                                                                               
      child?.layout(constraints);                                                                                                                                                                                                                           
      return;                                                                                                                                                                                                                                               
    }                                                                                                                                                                                                                                                       
```

File: packages/flutter/lib/src/rendering/box.dart (hunk 4 above)
Comment (Hixie) [inline packages/flutter/lib/src/rendering/box.dart]: It would make the later text confusing ("this second render object" wouldn't be as obvious), so I've left it as object.

File: packages/flutter/lib/src/rendering/box.dart (hunk 5 above)
Comment (Hixie) [inline packages/flutter/lib/src/rendering/box.dart]: fixed throughout the file

File: packages/flutter/lib/src/widgets/animated_cross_fade.dart (hunk 6 above)
Comment (Hixie) [inline packages/flutter/lib/src/widgets/animated_cross_fade.dart]: added your suggested text.

Comment (Hixie) [inline packages/flutter/lib/src/widgets/animated_cross_fade.dart]: ~~added your suggested text~~

Actually, i've just put in some sample code instead.

**Comment by Hixie** [comment]
updated. will land on green, if it ever goes green.
